ray_on_golem/server/services/golem/golem.py

```diff
@@ -343,51 +345,139 @@ def _score_with_provider_data(
+        await context.deploy(deploy_args, timeout=timedelta(minutes=5))
+
+        logger.info(f"Starting `{context._activity}`")
+        await context.start()
+
+    async def _upload_node_configuration(
+        self, context: WorkContext, ip: str, ssh_public_key_data: str
+    ):
+        logger.info(f"Running initial commands on `{context._activity}`")
         hostname = ip.replace(".", "-")
-        batch = await context.create_batch()
-        batch.deploy(deploy_args)
-        batch.start()
-        batch.run("echo 'ON_GOLEM_NETWORK=1' >> /etc/environment")
-        batch.run(f"hostname '{hostname}'")
-        batch.run(f"echo '{hostname}' > /etc/hostname")
-        batch.run(f"echo '{ip} {hostname}' >> /etc/hosts")
-        batch.run("mkdir -p /root/.ssh")
-        batch.run(f'echo "{ssh_public_key_data}" >> /root/.ssh/authorized_keys')
-        batch.run("service ssh start")
-        await batch()
+        await context.run("echo 'ON_GOLEM_NETWORK=1' >> /etc/environment")
+        await context.run(f"hostname '{hostname}'")
+        await context.run(f"echo '{hostname}' > /etc/hostname")
+        await context.run(f"echo '{ip} {hostname}' >> /etc/hosts")
+        await context.run("mkdir -p /root/.ssh")
+        await context.run(f'echo "{ssh_public_key_data}" >> /root/.ssh/authorized_keys')
+
+    async def _start_ssh_server(self, context: WorkContext):
+        logger.info(f"Starting ssh service on `{context._activity}`")
+        await context.run("service ssh start")
+
+    async def _verify_ssh_connection(
+        self,
+        activity: Activity,
+        ip: str,
+        ssh_proxy_command: str,
+        ssh_user: str,
+        ssh_private_key_path: Path,
+    ) -> None:
+        ssh_command = self._get_ssh_command(ip, ssh_proxy_command, ssh_user, ssh_private_key_path)
+
+        logger.debug(f"Verifying ssh connection with:\n{ssh_command} uptime")
+
+        process = await asyncio.create_subprocess_shell(
+            f"{ssh_command} uptime",
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+        )
+
+        stdout, stderr = await process.communicate()
+
+        msg = f"Validation of ssh connection done on {activity}:\n[cmd exited with {process.returncode}]"
+        if stdout:
+            msg += f" [stdout] {stdout.decode().strip()}"
+        if stderr:
+            msg += f" [stderr] {stderr.decode().strip()}"
+
+        if process.returncode != 0:
+            logger.warning(msg)
+            raise Exception(msg)
+
+        logger.info(msg)
+
+    def _get_ssh_command(
+        self, ip: str, ssh_proxy_command: str, ssh_user: str, ssh_private_key_path: Path
+    ) -> str:
+        return (
+            "ssh "
+            "-o StrictHostKeyChecking=no "
+            "-o UserKnownHostsFile=/dev/null "
+            f'-o "ProxyCommand={ssh_proxy_command}" '
+            f"-i {ssh_private_key_path} "
+            f"{ssh_user}@{ip}"
+        )
```

Comment:
This function is nearly duplicated by `RayService._print_ssh_command`. It would be good to keep everything in one place.

Reply:
Done